### PR TITLE
21279 add involuntary dissolutions warning 

### PR DESCRIPTION
**Pull request by kzdev420**
*Issue #:* /bcgov/entity#21279

*Description of changes:*

### 1. Business that is eligible for involuntary dissolution but does not have process initiated yet
```
"warnings": [
            {
                "code": "MULTIPLE_ANNUAL_REPORTS_NOT_FILED",
                "message": "Multiple annual reports not filed.  Eligible for involuntary dissolution.",
                "warningType": "NOT_IN_GOOD_STANDING"
            }          
        ],
```
### 2. Business that is in the process of involuntary dissolution
```
"warnings": [
      {
        "code": "MULTIPLE_ANNUAL_REPORTS_NOT_FILED",
        "message": "Multiple annual reports not filed.  Eligible for involuntary dissolution.",
        "warningType": "NOT_IN_GOOD_STANDING"
      },
      {
        "code": "DISSOLUTION_IN_PROGRESS",
        "data": {
          "missingARs": 5,
          "missingTransitionFilings": false,
          "targetDissolutionDate": "2025-02-01",
          "warningsSent": 2
        },
        "message": "Business is in the process of involuntary dissolution.",
        "warningType": "INVOLUNTARY_DISSOLUTION"
      }
    ]
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/src/legal_api/models/business.py [hunk 1]
```diff
@@ -416,6 +417,18 @@ def good_standing(self):
...
+    def is_involuntary_dissolution(self):
+        """Return true if in dissolution, otherwise false."""
+        # check a business has a batch_processing entry that matches business_id and status is not COMPLETED
+        is_involuntary_dissolution = db.session.query(BatchProcessing, Batch).\
+            filter(BatchProcessing.business_id == self.id).\
+            filter(Batch.id == BatchProcessing.batch_id).\
+            filter(Batch.status != Batch.BatchStatus.COMPLETED).\
+            filter(Batch.batch_type == Batch.BatchType.INVOLUNTARY_DISSOLUTION).\
+            one_or_none()
+        return is_involuntary_dissolution is not None
```
[the remark this answers is not in the document]
Comment (kzdev420) [reply legal-api/src/legal_api/models/business.py:430]: I think `in_dissolution` is bool and returns true when it is in only process.
But, warnings are 2 cases. So, I thought like these
1. Business that is eligible for involuntary dissolution but does not have process initiated yet - `is_involuntary_dissolution`
2. Business that is in the process of involuntary dissolution - `in_dissolution`

File: legal-api/src/legal_api/services/warnings/warning.py [hunk 2]
```diff
@@ -25,5 +25,6 @@ def check_warnings(business: Business) -> list:
     # Currently only checks for missing business info warnings but in future other warning checks can be included
     # e.g. compliance checks - result.extend(check_compliance(business))
     result.extend(check_business(business))
+    result.extend(check_involuntary_dissolution(business))
```
[the remark this answers is not in the document]
Comment (kzdev420) [reply legal-api/src/legal_api/services/warnings/warning.py:28]:
```
eligible_types = [
            Business.LegalTypes.COMP.value,
            Business.LegalTypes.BC_ULC_COMPANY.value,
            Business.LegalTypes.BC_CCC.value,
            Business.LegalTypes.BCOMP.value,
            Business.LegalTypes.CONTINUE_IN.value,
            Business.LegalTypes.ULC_CONTINUE_IN.value,
            Business.LegalTypes.CCC_CONTINUE_IN.value,
            Business.LegalTypes.BCOMP_CONTINUE_IN.value,
            Business.LegalTypes.EXTRA_PRO_A.value,
            Business.LegalTypes.LIMITED_CO.value
        ]
```
I think we need to check all these business types.
But, currently, there are only 4 types (BC_CCC, BC_ULC_COMPANY, COMP, BCOMP) in corps_check.
So, I did like that

[the remark this answers is not in the document]
Comment (kzdev420) [reply legal-api/src/legal_api/services/warnings/warning.py]: Updated

File: legal-api/src/legal_api/models/business.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kzdev420) [reply legal-api/src/legal_api/models/business.py]: Updated

File: legal-api/src/legal_api/models/business.py [hunk 3]
```diff
@@ -132,6 +132,19 @@ class LegalTypes(str, Enum):
...
+        LegalTypes.BC_ULC_COMPANY.value,
+        LegalTypes.BC_CCC.value,
+        LegalTypes.BCOMP.value,
+        LegalTypes.CONTINUE_IN.value,
+        LegalTypes.ULC_CONTINUE_IN.value,
+        LegalTypes.CCC_CONTINUE_IN.value,
+        LegalTypes.BCOMP_CONTINUE_IN.value,
+        LegalTypes.EXTRA_PRO_A.value,
+        LegalTypes.LIMITED_CO.value
+    ]
```
Comment (argush3) [inline legal-api/src/legal_api/models/business.py:146]: you can use [CORPS](https://github.com/bcgov/lear/blob/acdc05ccff4e56131790e1cc650b42df95babfc9/legal-api/src/legal_api/models/business.py#L124-L133) in the business model instead.

File: legal-api/src/legal_api/services/warnings/business/business_checks/business.py [hunk 4]
```diff
@@ -28,11 +29,14 @@ def check_business(business: any) -> list:
...
-        (Business.LegalTypes.BC_CCC,
+        (Business.LegalTypes.BC_CCC.value,
          Business.LegalTypes.BC_ULC_COMPANY.value,
          Business.LegalTypes.COMP.value,
          Business.LegalTypes.BCOMP.value
          ):
         result = corps_check(business)
 
+    if business.legal_type in Business.ELIGIBLE_TYPES:
+        result.extend(involuntary_dissolution_check(business))
```
Comment (argush3) [inline legal-api/src/legal_api/services/warnings/business/business_checks/business.py:40]: please move into `corps_check`

File: legal-api/src/legal_api/services/warnings/business/business_checks/business.py [hunk 5]
```diff
@@ -28,11 +29,14 @@ def check_business(business: any) -> list:
              Business.LegalTypes.PARTNERSHIP.value):
         result = firms_check(business)
     elif business.legal_type in \
-        (Business.LegalTypes.BC_CCC,
+        (Business.LegalTypes.BC_CCC.value,
          Business.LegalTypes.BC_ULC_COMPANY.value,
          Business.LegalTypes.COMP.value,
          Business.LegalTypes.BCOMP.value
          ):
```
Comment (argush3) [inline legal-api/src/legal_api/services/warnings/business/business_checks/business.py:36]: can replace this corp list with [CORPS](https://github.com/bcgov/lear/blob/acdc05ccff4e56131790e1cc650b42df95babfc9/legal-api/src/legal_api/models/business.py#L124-L133) list in business model.

File: legal-api/src/legal_api/models/business.py (hunk 3 above)
Comment (argush3) [reply legal-api/src/legal_api/models/business.py:146]: Actually, `CORPS` is missing `LegalTypes.EXTRA_PRO_A.value` and `LegalTypes.LIMITED_CO.value`.

I think you should create an `ELIGIBLE_TYPES` list in the involuntary dissolutions service class.  Don't think it belongs in the business model.

Comment (kzdev420) [reply legal-api/src/legal_api/models/business.py]: Updated

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+def check_involuntary_dissolution(business: any) -> list:
+    """Check involuntary dissolution for warnings."""
+    result = []
+
+    if business.is_involuntary_dissolution:
+        result.append({
+            'code': BusinessWarningCodes.MULTIPLE_ANNUAL_REPORTS_NOT_FILED,
+            'message': 'Multiple annual reports not filed.  Eligible for involuntary dissolution.',
+            'warningType': WarningType.NOT_IN_GOOD_STANDING
+        })
```
[the remark this answers is not in the document]
Comment (argush3) [reply legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py:30]: As per our discussion, I've updated the design to support the `code` = "TRANSITION_NOT_FILED" as well for `warningType` = "NOT_IN_GOOD_STANDING".  

```
{
    "business": {
        "legalName": "XYZ TEST CORP.",
        "legalType": "BC",
        "state": "ACTIVE",
        "goodStanding": false,
        "warnings": [
            {
                "code": "<MULTIPLE_ANNUAL_REPORTS_NOT_FILED> or <TRANSITION_NOT_FILED>",
                "message": "<Multiple annual reports not filed> or <Transition filing not filed>.  Eligible for involuntary dissolution.",
                "warningType": "NOT_IN_GOOD_STANDING"
            }          
        ],
        ...
    }
}
```

`check_business_eligibility` has been updated to pass back eligibility details so you can distinguish between  ARs overdue or missing transition filing when a business is not in good standing.

[the remark this answers is not in the document]
Comment (kzdev420) [reply legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py]: updated

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        'message': 'Transition filing not filed.  Eligible for involuntary dissolution.',
+        'warningType': WarningType.NOT_IN_GOOD_STANDING
+    }
+
+    eligibility, details = InvoluntaryDissolutionService.check_business_eligibility(business.identifier, True)
+    if eligibility:
+        if details.ar_overdue:
+            result.append(ar_overdue_warning)
+        if details.transition_overdue:
+            result.append(transition_warning)
```
Comment (argush3) [inline legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py:41]: if `details.transition_overdue` is True then we should just append that.  We only dissolve a business for one of these reasons.  Transition takes precedence.

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    elif business.in_dissolution:
+        _, dis_details = InvoluntaryDissolutionService.check_business_eligibility(business.identifier, False)
+        if dis_details.ar_overdue:
+            result.append(ar_overdue_warning)
+        if dis_details.transition_overdue:
+            result.append(transition_warning)
```
Comment (argush3) [inline legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py:47]: same comment as above

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        batch_processing = business.batch_processing.one_or_none()
```
Comment (argush3) [inline legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py:48]: I think this needs to be a more precise query.  The batch tables could be used for other things in the future.

I think you should create a function in `InvoluntaryDissolutionsService` that does what `business.in_dissolution` essentially does.

``` python
@staticmethod
    def get_in_dissolution_batch_processing(business_id: int):
        """Fetch the BatchProcessing record for a business that is in the process of involuntary dissolution."""
        return db.session.query(BatchProcessing, Batch).\
            filter(BatchProcessing.business_id == business_id).\
            filter(BatchProcessing.status.notin_([BatchProcessing.BatchProcessingStatus.COMPLETED,
                                                  BatchProcessing.BatchProcessingStatus.WITHDRAWN])). \
            filter(Batch.id == BatchProcessing.batch_id).\
            filter(Batch.status != Batch.BatchStatus.COMPLETED).\
            filter(Batch.batch_type == Batch.BatchType.INVOLUNTARY_DISSOLUTION).\
            one_or_none()
```

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Service to check involuntary dissolution for a business."""
+from legal_api.services.involuntary_dissolution import InvoluntaryDissolutionService
+
+from . import BusinessWarningCodes, WarningType
+
+
+def check_business(business: any) -> list:
```
Comment (argush3) [inline legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py:21]: This isn't the legal name work where we had different business classes. 

It should be `def check_business(business: Business) -> list:`

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+)
+from legal_api.services.warnings.business.business_checks import WarningType
+from legal_api.services.warnings.business.business_checks.involuntary_dissolution import check_business
+
+@pytest.mark.parametrize('test_name, no_dissolution, batch_status, batch_processing_status', [
+    ('NOT_ELIGIBLE', True, None, None),
+    ('ELIGIBLE', False, 'PROCESSING', 'COMPLETED'),
+    ('IN_DISSOLUTION', False, 'PROCESSING', 'PROCESSING')
+])
+def test_check_business(session, test_name, no_dissolution, batch_status, batch_processing_status):
```
Comment (argush3) [inline legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py:31]: can you make sure this test, tests overdueTransition = True scenario too?

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py (hunk 10 above)
Comment (kzdev420) [reply legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py]: Updated

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py (hunk 8 above)
Comment (kzdev420) [reply legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py]: Updated

File: legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py (hunk 7 above)
Comment (kzdev420) [reply legal-api/src/legal_api/services/warnings/business/business_checks/involuntary_dissolution.py]: Updated

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py (hunk 11 above)
Comment (argush3) [reply legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py:31]: as per our conversation, adding this scenario can be done in a separate PR

Comment (kzdev420) [reply legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py]: Updated

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+from datedelta import datedelta
+from registry_schemas.example_data import FILING_HEADER, RESTORATION, TRANSITION_FILING_TEMPLATE
+
+RESTORATION_FILING = copy.deepcopy(FILING_HEADER)
+RESTORATION_FILING['filing']['restoration'] = RESTORATION
+
+@pytest.mark.parametrize('test_name, no_dissolution, batch_status, batch_processing_status', [
+    ('NOT_ELIGIBLE', True, None, None),
+    ('ELIGIBLE', False, 'PROCESSING', 'COMPLETED'),
```
Comment (argush3) [inline legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py:38]: can you rename this test to `ELIGIBLE_AR_OVERDUE`?

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    ('ELIGIBLE_TRANSITION_OVERDUE', False, 'PROCESSING', 'COMPLETED'),
+    ('IN_DISSOLUTION', False, 'PROCESSING', 'PROCESSING'),
```
Comment (argush3) [inline legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py:40]: can you rename this test to IN_DISSOLUTION_AR_OVERDUE?

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py (hunk 12 above)
Comment (kzdev420) [reply legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py]: Updated

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py (hunk 13 above)
Comment (kzdev420) [reply legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py]: Updated

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        factory_completed_filing(business, RESTORATION_FILING, filing_type='restoration', filing_date=effective_date)
+
+    if not no_dissolution:
+        batch = factory_batch(
+            batch_type = Batch.BatchType.INVOLUNTARY_DISSOLUTION,
+            status = batch_status,
+        )
+        meta_data = {
+            'overdueARs': False,
+            'overdueTransition': True,
```
Comment (argush3) [inline legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py:57]: this needs to change depending on the scenarios

Comment (kzdev420) [reply legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py]: Updated

**Review by JazzarKarim — APPROVED** [review]

File: legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+    if test_name == 'NOT_ELIGIBLE':
+        assert len(result) == 0
+    else:
+        if test_name.startswith('IN_DISSOLUTION'):
+            assert len(result) == 2
+            assert result[1]['code'] == 'DISSOLUTION_IN_PROGRESS'
+            assert result[1]['message'] == 'Business is in the process of involuntary dissolution.'
+            assert result[1]['warningType'] == WarningType.INVOLUNTARY_DISSOLUTION
+            assert result[1]['data'] == json.dumps(meta_data)
```
Comment (argush3) [inline legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py:81]: Can you add an explicit assertion for the expected values for `overdueARs` and `overdueTransition` depending on the scenario.

Comment (kzdev420) [reply legal-api/tests/unit/services/warnings/business/business_checks/test_involuntary_dissolution.py:81]: Added

**Comment by argush3** [comment]
As per our conversation, the 2 broken tests can be fixed in another PR.

image

**Review by argush3 — APPROVED** [review]

**Pull request merged**